2022-10-27 14:59:13.497 | INFO     | sc2.protocol:_execute:72 - Client status changed to Status.launched (was None)
2022-10-27 14:59:13.499 | INFO     | sc2.controller:create_game:37 - Creating new game
2022-10-27 14:59:13.500 | INFO     | sc2.controller:create_game:38 - Map:     2000AtmospheresAIE
2022-10-27 14:59:13.501 | INFO     | sc2.controller:create_game:39 - Players: Bot Turtles(Terran), Computer VeryHard(Random, RandomBuild)
2022-10-27 14:59:13.504 | INFO     | sc2.protocol:_execute:72 - Client status changed to Status.init_game (was Status.launched)
2022-10-27 14:59:23.653 | INFO     | sc2.protocol:_execute:72 - Client status changed to Status.in_game (was None)
2022-10-27 14:59:23.656 | INFO     | sc2.main:_play_game:221 - Player 1 - Bot Turtles(Terran)
2022-10-27 14:59:24.497 | 00:00    0ms    0M    0G  12/ 15U [Version] 5.0.10.88500
2022-10-27 14:59:24.498 | 00:00    0ms    0M    0G  12/ 15U [Version] 5.0.6 Balance mod recognized.
2022-10-27 14:59:24.579 | 00:00    0ms    0M    0G  12/ 15U [ZoneManager] Map set to: MapName.Atmospheres2000 from name: 2000 Atmospheres AIE and hash: 3307847.
2022-10-27 14:59:25.617 | ERROR    | sc2.main:initialize_first_step:140 - Caught unknown exception in AI on_start: 'MacroBio' object has no attribute 'ai'
Traceback (most recent call last):

  File "C:\Users\sword\OneDrive\Desktop\27turtles\run.py", line 18, in <module>
    main()
    └ <function main at 0x000001FAB43E55A0>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\run.py", line 13, in main
    result, opponentid = run_ladder_game(bot)
                         │               └ <sc2.player.Bot object at 0x000001FAB4416740>
                         └ <function run_ladder_game at 0x000001FA91D09360>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\ladder.py", line 39, in run_ladder_game
    return stand_alone_game(bot), None
           │                └ <sc2.player.Bot object at 0x000001FAB4416740>
           └ <function stand_alone_game at 0x000001FAB1CA95A0>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\ladder.py", line 168, in stand_alone_game
    return sc2.main.run_game(sc2.maps.get(map_name), [bot, Computer(Race.Random, Difficulty.VeryHard)], realtime=False,)
           │   │    │        │   │    │   │           │    │        │    │       │          └ <Difficulty.VeryHard: 7>
           │   │    │        │   │    │   │           │    │        │    │       └ <enum 'Difficulty'>
           │   │    │        │   │    │   │           │    │        │    └ <Race.Random: 4>
           │   │    │        │   │    │   │           │    │        └ <enum 'Race'>
           │   │    │        │   │    │   │           │    └ <class 'sc2.player.Computer'>
           │   │    │        │   │    │   │           └ <sc2.player.Bot object at 0x000001FAB4416740>
           │   │    │        │   │    │   └ '2000AtmospheresAIE'
           │   │    │        │   │    └ <function get at 0x000001FAB1D584C0>
           │   │    │        │   └ <module 'sc2.maps' from 'C:\\Users\\sword\\OneDrive\\Desktop\\27turtles\\sc2\\maps.py'>
           │   │    │        └ <module 'sc2' from 'C:\\Users\\sword\\OneDrive\\Desktop\\27turtles\\sc2\\__init__.py'>
           │   │    └ <function run_game at 0x000001FAB1E38DC0>
           │   └ <module 'sc2.main' from 'C:\\Users\\sword\\OneDrive\\Desktop\\27turtles\\sc2\\main.py'>
           └ <module 'sc2' from 'C:\\Users\\sword\\OneDrive\\Desktop\\27turtles\\sc2\\__init__.py'>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\sc2\main.py", line 496, in run_game
    result: Result = asyncio.run(_host_game(map_settings, players, **kwargs))
                     │       │   │          │             │          └ {'realtime': False}
                     │       │   │          │             └ [<sc2.player.Bot object at 0x000001FAB4416740>, <sc2.player.Computer object at 0x000001FAB4417580>]
                     │       │   │          └ Map(C:\Program Files (x86)\StarCraft II\maps\2000AtmospheresAIE.SC2Map)
                     │       │   └ <function _host_game at 0x000001FAB1E389D0>
                     │       └ <function run at 0x000001FA939CD7E0>
                     └ <module 'asyncio' from 'C:\\Users\\sword\\AppData\\Local\\Programs\\Python\\Python310\\lib\\asyncio\\__init__.py'>

  File "C:\Users\sword\AppData\Local\Programs\Python\Python310\lib\asyncio\runners.py", line 44, in run
    return loop.run_until_complete(main)
           │    │                  └ <coroutine object _host_game at 0x000001FAB4396E30>
           │    └ <function BaseEventLoop.run_until_complete at 0x000001FA939CF1C0>
           └ <ProactorEventLoop running=True closed=False debug=False>

  File "C:\Users\sword\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 633, in run_until_complete
    self.run_forever()
    │    └ <function ProactorEventLoop.run_forever at 0x000001FA93A7BBE0>
    └ <ProactorEventLoop running=True closed=False debug=False>

  File "C:\Users\sword\AppData\Local\Programs\Python\Python310\lib\asyncio\windows_events.py", line 321, in run_forever
    super().run_forever()

  File "C:\Users\sword\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 600, in run_forever
    self._run_once()
    │    └ <function BaseEventLoop._run_once at 0x000001FA939D0CA0>
    └ <ProactorEventLoop running=True closed=False debug=False>

  File "C:\Users\sword\AppData\Local\Programs\Python\Python310\lib\asyncio\base_events.py", line 1896, in _run_once
    handle._run()
    │      └ <function Handle._run at 0x000001FA9392C280>
    └ <Handle Task.task_wakeup(<Future finished result=None>)>

  File "C:\Users\sword\AppData\Local\Programs\Python\Python310\lib\asyncio\events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
    │    │            │    │           │    └ <member '_args' of 'Handle' objects>
    │    │            │    │           └ <Handle Task.task_wakeup(<Future finished result=None>)>
    │    │            │    └ <member '_callback' of 'Handle' objects>
    │    │            └ <Handle Task.task_wakeup(<Future finished result=None>)>
    │    └ <member '_context' of 'Handle' objects>
    └ <Handle Task.task_wakeup(<Future finished result=None>)>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\sc2\main.py", line 369, in _host_game
    result = await _play_game(players[0], client, realtime, portconfig, game_time_limit, rgb_render_config)
                   │          │           │       │         │           │                └ None
                   │          │           │       │         │           └ None
                   │          │           │       │         └ None
                   │          │           │       └ False
                   │          │           └ <sc2.client.Client object at 0x000001FAB44150C0>
                   │          └ [<sc2.player.Bot object at 0x000001FAB4416740>, <sc2.player.Computer object at 0x000001FAB4417580>]
                   └ <function _play_game at 0x000001FAB1E38820>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\sc2\main.py", line 226, in _play_game
    result = await _play_game_ai(client, player_id, player.ai, realtime, game_time_limit)
                   │             │       │          │      │   │         └ None
                   │             │       │          │      │   └ False
                   │             │       │          │      └ <turtles.bot.Turtles object at 0x000001FA939C6C80>
                   │             │       │          └ <sc2.player.Bot object at 0x000001FAB4416740>
                   │             │       └ 1
                   │             └ <sc2.client.Client object at 0x000001FAB44150C0>
                   └ <function _play_game_ai at 0x000001FAB1E38670>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\sc2\main.py", line 145, in _play_game_ai
    result = await initialize_first_step()
                   └ <function _play_game_ai.<locals>.initialize_first_step at 0x000001FAB43E63B0>

> File "C:\Users\sword\OneDrive\Desktop\27turtles\sc2\main.py", line 136, in initialize_first_step
    await ai.on_start()
          │  └ <function KnowledgeBot.on_start at 0x000001FAB4256C20>
          └ <turtles.bot.Turtles object at 0x000001FA939C6C80>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\sharpy\knowledges\knowledge_bot.py", line 71, in on_start
    await self.knowledge.start()
          │    │         └ <function Knowledge.start at 0x000001FAB1EBAF80>
          │    └ <sharpy.knowledges.knowledge.Knowledge object at 0x000001FA91996590>
          └ <turtles.bot.Turtles object at 0x000001FA939C6C80>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\sharpy\knowledges\knowledge.py", line 151, in start
    await manager.post_start()
          │       └ <function ActManager.post_start at 0x000001FAB1D20F70>
          └ <sharpy.managers.core.act_manager.ActManager object at 0x000001FAB4517FA0>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\sharpy\managers\core\act_manager.py", line 25, in post_start
    self._act = await self._act_or_func()
    │                 │    └ <bound method Turtles.create_plan of <turtles.bot.Turtles object at 0x000001FA939C6C80>>
    │                 └ <sharpy.managers.core.act_manager.ActManager object at 0x000001FAB4517FA0>
    └ <sharpy.managers.core.act_manager.ActManager object at 0x000001FAB4517FA0>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\turtles\bot.py", line 48, in create_plan
    self.build = MacroBio()
    │    │       └ <class 'turtles.biobuild.MacroBio'>
    │    └ <function BotAI.build at 0x000001FAB1C2BE20>
    └ <turtles.bot.Turtles object at 0x000001FA939C6C80>

  File "C:\Users\sword\OneDrive\Desktop\27turtles\turtles\biobuild.py", line 226, in __init__
    Step(self.ai.supply_army > 22,NoRallyAttack(22),skip=self.lost_units_manager.own_lost_type(UnitTypeId.MARINE) > 15)
    │    │                        │                      │                                     │          └ UnitTypeId.MARINE
    │    │                        │                      │                                     └ <enum 'UnitTypeId'>
    │    │                        │                      └ <turtles.biobuild.MacroBio object at 0x000001FAB6093850>
    │    │                        └ <class 'turtles.combat.norallyattack.NoRallyAttack'>
    │    └ <turtles.biobuild.MacroBio object at 0x000001FAB6093850>
    └ <class 'sharpy.plans.build_step.Step'>

AttributeError: 'MacroBio' object has no attribute 'ai'
2022-10-27 14:59:25.746 | ERROR    | sc2.main:initialize_first_step:141 - Resigning due to previous error
2022-10-27 14:59:25.747 | Result: Defeat
2022-10-27 14:59:25.748 | Duration: 00:00
2022-10-27 14:59:25.749 | Step time avg: 0
2022-10-27 14:59:25.749 | Step time max: 0
2022-10-27 14:59:25.750 | [LostUnitsContents] My lost units minerals and gas: (0, 0)
2022-10-27 14:59:25.751 | [LostUnitsContents] My lost units minerals by score: 0.0
2022-10-27 14:59:25.751 | [LostUnitsContents] Enemy lost units minerals and gas: (0, 0)
2022-10-27 14:59:25.752 | [LostUnitsContents] Enemy lost units minerals by score: 0.0
2022-10-27 14:59:25.755 | [GameAnalyzerEnd] Own units:
2022-10-27 14:59:25.756 | [GameAnalyzerEnd] Enemy units:
2022-10-27 14:59:25.756 | [GameAnalyzerEnd] Minerals max 0 Average 0
2022-10-27 14:59:25.757 | [GameAnalyzerEnd] Vespene max 0 Average 0
2022-10-27 14:59:25.758 | INFO     | sc2.main:_play_game:228 - Result for player 1 - Bot Turtles(Terran): Defeat
2022-10-27 14:59:26.787 | INFO     | sc2.protocol:_execute:72 - Client status changed to Status.launched (was Status.in_game)
2022-10-27 14:59:26.790 | INFO     | sc2.protocol:_execute:72 - Client status changed to Status.quit (was Status.launched)
2022-10-27 14:59:26.791 | INFO     | sc2.sc2process:_close_connection:231 - Closing connection at 63114...
2022-10-27 14:59:26.793 | INFO     | sc2.sc2process:kill_all:33 - kill_switch: Process cleanup for 1 processes
